src/ray/core_worker/pubsub/pubsub_coordinator.cc

```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/pubsub/pubsub_coordinator.h"
+
+namespace ray {
+
+void PubsubCoordinator::Connect(const NodeID &subscriber_node_id,
+                                LongPollConnectCallback long_poll_connect_callback) {
+  RAY_LOG(DEBUG) << "Long polling connection is initiated by " << subscriber_node_id;
+  if (connection_pool_.count(subscriber_node_id) == 0) {
+    RAY_CHECK(
+        connection_pool_.emplace(subscriber_node_id, long_poll_connect_callback).second);
+  }
+}
+
+void PubsubCoordinator::RegisterSubscriber(const NodeID &subscriber_node_id,
+                                           const ObjectID &object_id) {
+  RAY_LOG(DEBUG) << "object id " << object_id << " is subscribed by "
+                 << subscriber_node_id;
+  auto &subscribers = objects_to_subscribers_[object_id];
+  subscribers.emplace(subscriber_node_id);
+}
+
+void PubsubCoordinator::Publish(const ObjectID &object_id,
+                                bool publish_message_if_possible) {
+  objects_to_publish_.emplace(object_id);
+  if (publish_message_if_possible) {
+    PublishAllMessages();
```

Comment:
Why do we need to loop over all the objects here? I'd imagine it would go like this:
for (subscriber : subscribers_of(object_id)) {
   add to mailbox of subscriber
   trigger rpc response to subscriber if a long poll is active
}

Reply:
I think it is just a slightly different implementation, but your suggestion sounds much cleaner.

Current impl: cache published objects until everything is published to subscribers. (and if long polling connection is not there, we just wait until it comes). 
New suggestion: cache published objects to the mailbox and reply when long polling is reinitiated.

I think your suggested implementation is cleaner, but it will require some special handling for this case;

- Object is added to the mailbox.
- Long polling connection wasn't there, so we wait.
- And then the subscriber fails.

In this case, we need to clean up subscribers when we get the node death message. I think this is also pretty simple, so let me follow this implementation.